### DISCUSSION TITLE
System Agent of Intel Gen 11th & 12th

### DISCUSSION THREAD

**Discussion by cyring** [cyring/CoreFreq#338] [Show and tell]
For testings in commit 2bfbd5bcbb3c0480f6abd6e3ae9fbddde561e0b9, the SA voltage

![2022-04-27-141202_972x550_scrot](https://user-images.githubusercontent.com/11563789/165515757-58869649-9d88-44f7-94a0-4efad61b73a7.png)

![2022-04-27-141228_644x550_scrot](https://user-images.githubusercontent.com/11563789/165515777-3106e380-a0a3-43ef-96df-0644681bc1b2.png)

* Please note, the voltage VID and value displayed in `SA` are fixed.

**Comment by BugReporterZ** [comment]
The first screenshot is while running an OpenGL animation in Blender, which makes SA show power of a few watts (but this is actually due to the iGPU):

![image](https://user-images.githubusercontent.com/26941368/165519953-54b96fb1-e181-401d-bc70-9e0479b6dd01.png)

The second screenshot is while stressing the two fastest cores:

![image](https://user-images.githubusercontent.com/26941368/165520235-09008958-8c97-4e30-878d-e07c788a3069.png)

I'm not sure what difference there is between VID and other voltages shown in CoreFreq. All of these voltages, if reported by the CPU via MSRs, may not necessarily accurately represent actual input from the voltage regulators. For that, one should check out external sensors from the motherboard, i.e. Vcore ("true" Vcore). 

On Windows, users/overclockers often treat VID (voltages reported by the CPU) and Vcore (CPU core voltage from the motherboard) as separate when monitoring their CPU, with only the latter giving reliable readings.

**Reply by BugReporterZ** [reply]
It seems that per-core reported voltages disappeared, by the way. They now all show 0.000V.

The core voltage requested by the CPU to the voltage regulators is always the highest requested by any core however, so per-core readings may not be that useful (at least yet, until fully-integrated voltage regulators/FIVR will allow the CPU to have truly different voltages for each core).

**Reply by cyring** [reply]
I will fix the Vcore regression.

This time, registers are sourced from the PCI space, not MSR.
VID, the raw bits read from registers.
VID through manufacturer's formula is then computed as a voltage.

Don't expect things from the Windows and BIOS world.
It's open-source with no Manufacturer support.

**Reply by cyring** [reply]
@BugReporterZ commit 3db4a1569317d5db7c6e057ea53e0fc85dd363c7 you should get the Vcore fix and btw the Time Window format fix

**Reply by BugReporterZ** [reply]
Core voltages now work, and I don't seem to see crazy values in Time window settings.

![image](https://user-images.githubusercontent.com/26941368/165545596-02c1a24a-6c93-410c-a2a0-dd67aa5b6607.png)

**Comment by cyring** [comment]
Nice to have are Registers about `Fused ratio limit`, `OC ratio limit`, `Voltage offset`, `Voltage/frequency curve`, `Target voltage`, `iccMax` within domains `IA`, `Ring`, `SA`, `GT` listed in the [screenshot](https://user-images.githubusercontent.com/26941368/166140463-0383d6cf-7565-4238-8ad9-e172fe0fd00b.png) nicely provided by @BugReporterZ

**Reply by BugReporterZ** [reply]
AC/DC loadlines are also part of voltage settings. In detail, AC loadline is a current-based voltage correction applied when Adaptive voltages are used; DC loadline affects reported power values. This is also in the CPU datasheet, although I'm not aware of related MSR documentation. Below are samples from the 11th gen CPU datasheet:

![image](https://user-images.githubusercontent.com/26941368/166155474-fdce38d8-63a3-4290-9915-cce48ee63bd9.png)

![image](https://user-images.githubusercontent.com/26941368/166155463-40908c9e-2c0d-428e-ad2e-a79509d63415.png)

**Reply by cyring** [reply]
What we have, at least, is the so called `SA Overclocking > Target voltage` of `1.0510 V` 😊
